opengovsg/plumber
# PLU-91: [SGID-LOGIN-3] **RFC** Implement SGID failure modal

## Problem
We want to support SGID login as an alternative to OTP.

## Solution
As opposed to the usual SGID approach of creating routes on the back end, we will implement SGID using front end routes + GraphQL instead (to avoid us mixing and matching classic REST vs GraphQL where possible).

### High Level Approach
_See #202 for full details._

1. Create a new "login with SGID" GraphQL mutation, which logs in users with a single work email, or return the list of work emails to the front end.
2. Create a new "login with selected SGID" mutation, which logs multi-hatted users in after they have selected their email.
3. Create a new login with SGID button, which sends the user to the SGID servers to start the login process
4. Add a new front end route, `/login/sgid/redirect`, which processes auth data from SGID + renders work email list for selection.

### PR Notes
This PR implements the failure page that users see if we don't find anything in POCDEX. It adds a `not_sgid_eligible` URL query param to the login page, and if it seems this param:
* Renders a failure message modal
* Disables the SGID login button

### RFC Needed
Design and UX comments? This has not been scrubbed by our PD yet :P

**SCREENSHOTS**
![Screenshot 2023-09-11 at 3 17 39 PM](https://github.com/opengovsg/plumber/assets/135598754/5c98cf25-644e-4b73-9027-cb2070938870)

![Screenshot 2023-09-11 at 3 19 45 PM](https://github.com/opengovsg/plumber/assets/135598754/3c46ec08-79fa-462a-b8d1-9b6028e229b2)

## Tests
- Check that modal shows up if I go to the login URL with this query param.
- Check that SGID login button is hidden if I go to the login URL with this query param.
- Simulate no POCDEX data in GraphQL, and check that I get redirected to this page.
- Regression test: check that modal does not appear, and SGID button is present, when I go to the login URL without this query param.
